chunky/src/java/se/llbit/math/BigPackedOctree.java

```diff
@@ -0,0 +1,773 @@
+package se.llbit.math;
+
+import org.apache.commons.math3.util.FastMath;
+import se.llbit.chunky.block.Air;
+import se.llbit.chunky.block.Block;
+import se.llbit.chunky.block.UnknownBlock;
+import se.llbit.chunky.block.Water;
+import se.llbit.chunky.chunk.BlockPalette;
+import se.llbit.chunky.model.TexturedBlockModel;
+import se.llbit.chunky.model.WaterModel;
+import se.llbit.chunky.renderer.scene.Scene;
+import se.llbit.chunky.world.Material;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+
+import static se.llbit.math.Octree.BRANCH_NODE;
+import static se.llbit.math.Octree.DATA_FLAG;
+
+/**
+ * This is a big packed representation of an octree
+ * the design is almost identical to the packed octree
+ * but uses several long arrays to have
+ */
+public class BigPackedOctree implements Octree.OctreeImplementation {
+  /**
+   * The design is almost the same that packed octree, the differences being:
+   *  - The data is split into several long arrays instead of a single int array
+   *  - The data for one node is store in one long instead of two ints, when the node
+   *    is a branch node, the long represents the index of the first child. When the
+   *    node is a leaf node, the upper 32 bits of the long contain the opposite of the node type
+   *    and the lower 32 bits are the node data (that way, for a leaf node the sign bit will be set
+   *    and the long can simply be compared with 0 to determine if it is a branch or a leaf)
+   *
+   */
+  private ArrayList<long[]> treeData = new ArrayList<>();
+
+  /**
+   * The max size of an array we allow is a bit less than the max value an integer can have
+   */
+  private static final int MAX_ARRAY_SIZE = 1 << 30; // MAX_INT is 2^31 - 1 but it is useful to use a power of 2 so we use the biggest power of 2 that is small that MAX_INT
+  private final long SUB_ARRAY_MASK = MAX_ARRAY_SIZE - 1;
+  private final long FULL_ARRAY_MASK = ~SUB_ARRAY_MASK;
```

Comment:
^-- these two could also be static final.

Reply:
Fixed